src/ctap/pin_protocol_v1.rs

```diff
@@ -0,0 +1,1228 @@
+// Copyright 2020 Google LLC
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+use super::command::AuthenticatorClientPinParameters;
+use super::data_formats::{ClientPinSubCommand, CoseKey, GetAssertionHmacSecretInput};
+use super::response::{AuthenticatorClientPinResponse, ResponseData};
+use super::status_code::Ctap2StatusCode;
+use super::storage::PersistentStore;
+#[cfg(feature = "with_ctap2_1")]
+use alloc::string::String;
+use alloc::vec::Vec;
+use core::convert::TryInto;
+use crypto::cbc::{cbc_decrypt, cbc_encrypt};
+use crypto::hmac::{hmac_256, verify_hmac_256_first_128bits};
+use crypto::rng256::Rng256;
+use crypto::sha256::Sha256;
+use crypto::Hash256;
+#[cfg(all(test, feature = "with_ctap2_1"))]
+use enum_iterator::IntoEnumIterator;
+use subtle::ConstantTimeEq;
+
+// Those constants have to be multiples of 16, the AES block size.
+pub const PIN_AUTH_LENGTH: usize = 16;
+const PIN_PADDED_LENGTH: usize = 64;
+const PIN_TOKEN_LENGTH: usize = 32;
+
+/// Checks the given pin_auth against the truncated output of HMAC-SHA256.
+/// Returns LEFT(HMAC(hmac_key, hmac_contents), 16) == pin_auth).
+fn verify_pin_auth(hmac_key: &[u8], hmac_contents: &[u8], pin_auth: &[u8]) -> bool {
+    if pin_auth.len() != PIN_AUTH_LENGTH {
+        return false;
+    }
+    verify_hmac_256_first_128bits::<Sha256>(
+        hmac_key,
+        hmac_contents,
+        array_ref![pin_auth, 0, PIN_AUTH_LENGTH],
+    )
+}
+
+/// Decrypts the HMAC secret salt(s) that were encrypted with the shared secret.
```

Comment:
This function does encrypt the HMAC-secret output. To compute those, it has to decrypt the salt(s) first. I restructured the comment to mention the most important parts of it first, instead of a chronological description.